### PR TITLE
[Spec Decode] Fix hidden-state extraction block size for hybrid verifiers

### PR DESCRIPTION
**Pull request by imargulis**
### Purpose
`extract_hidden_states` + `ExampleHiddenStatesConnector` silently produces
**all-zero hidden states** on hybrid (Mamba + attention) verifiers such as
`Qwen3.5-35B-A3B`, and crashes with a CUDA `index out of bounds` at scale.
Training on the zero/garbage states "converges" (validation looks fine) but the
resulting draft model has ~0% real acceptance — see vllm-project/speculators#613.
This supersedes #44328, narrowed to the actual root cause on current `main`.
### Root cause
Hybrid support was added in #39949, where the connector read `slot_mapping`
directly from `attn_metadata` (write-slots == read-slots, correct by
construction). The async-write rework (#43805) moved extraction to
`get_finished()`, where `attn_metadata` is gone, so it now **rebuilds**
`slot_mapping` from `block_ids × self._block_size` with
`self._block_size = cache_config.block_size`.
For hybrid models vLLM bumps the global `cache_config.block_size` to a common
multiple, while the page-aligned hidden-states KV cache group keeps its own
(smaller) block size. Concrete example for `Qwen3.5-35B-A3B`: global block size `528`,
hidden-states buffer block size `22` (`= 528/24`). Rebuilding read slots with
`528` while the buffer/decode use `22` reads block `b·24` instead of `b` →
unwritten (zero) slots, and `b·24` eventually exceeds `num_blocks` → the OOB.
#45849 fixed group detection + a NaN (block tracking) for hybrid, but did **not**
touch the block size, so the all-zeros remained.
### Changes
`vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py`
- Identify the hidden-states group by `HiddenStateCacheSpec` **type**
  (`_find_cache_kv_group_id`) instead of `"cache_only_layers"` layer-name
  substrings — robust on both scheduler and worker side.
- Derive the connector block size from that group's spec
  (`_get_cache_block_size`) instead of `cache_config.block_size`.
- Assert the derived block size matches the registered buffer in
  `register_kv_caches` (`self._block_size == self._kv_cache.shape[1]`) so this
  class of mismatch fails loudly instead of corrupting silently.
`vllm/model_executor/models/extract_hidden_states.py`
- `basic_cache` ignores padding slots (`slot_mapping < 0`) so padded /
  chunked-prefill batches don't scatter hidden states into block 0.
### Test Plan
Isolated hidden-state extraction (no training) on `Qwen/Qwen3.5-35B-A3B`,

```
#!/bin/bash
# Isolate vLLM hidden-state EXTRACTION from speculator training.
#
# Launches a vLLM hidden-states-extraction server for a (possibly hybrid)
# verifier model, then submits a small number of single requests and validates
# the extracted hidden states -- covering both the ONLINE client path
# (scripts/extract_hidden_states_smoke.py, the same call train.py makes) and the
# OFFLINE path (scripts/data_generation_offline.py).
#
# Goal: confirm hidden-state extraction works for a given verifier (e.g. the
# hybrid Qwen3.5/3.6-A3B models) independent of training, or pinpoint exactly
# where it breaks.
#
# Usage:
#   bash examples/train/extract_hidden_states_smoke.sh
# Reproduce a clean traceback on failure:
#   CUDA_LAUNCH_BLOCKING=1 TORCHDYNAMO_DISABLE=1 \
#     bash examples/train/extract_hidden_states_smoke.sh
set -euo pipefail

# ============ Configuration ============
MODEL="Qwen/Qwen3.5-35B-A3B"
DATA="./output/dflash_qwen3_5_35b_a3b_ultrachat"   # produced by prepare_data.py
TARGET_LAYER_IDS="2 11 20 29 38"                   # last layer auto-appended
VLLM_PORT=8000
MAX_MODEL_LEN=32768

NUM_SAMPLES=200            # bump this (e.g. 200) to chase multi-request issues
START_INDEX=0
REQUEST_MODE="both"     # online | offline | both

VLLM_GPUS="0,1,2,3"
# Verifier parallelism. TP=4/DP=1 = single engine sharded over 4 GPUs and
# gives the cleanest tracebacks. Set TP=1/DP=4 to mirror the training launcher.
TP_SIZE=4
DP_SIZE=1
ENFORCE_EAGER=0         # 1 => --enforce-eager (clean tracebacks); 0 => default
# =======================================

VLLM_LOG="$DATA/vllm_extract_smoke.log"
OFFLINE_OUT="$DATA/hs_smoke_offline"
mkdir -p "$DATA"

EAGER_FLAG=""
[ "$ENFORCE_EAGER" = "1" ] && EAGER_FLAG="--enforce-eager"

echo "=== Launching vLLM (hidden-states extraction) -> $VLLM_LOG ==="
CUDA_LAUNCH_BLOCKING="${CUDA_LAUNCH_BLOCKING:-1}" CUDA_VISIBLE_DEVICES="$VLLM_GPUS" \
  python scripts/launch_vllm.py "$MODEL" \
    --target-layer-ids $TARGET_LAYER_IDS \
    -- --tensor-parallel-size "$TP_SIZE" --data-parallel-size "$DP_SIZE" \
       --port "$VLLM_PORT" --max-model-len "$MAX_MODEL_LEN" $EAGER_FLAG \
  > "$VLLM_LOG" 2>&1 &
VLLM_PID=$!

cleanup() {
    echo "Stopping vLLM server ($VLLM_PID)..."
    kill "$VLLM_PID" 2>/dev/null || true
    wait "$VLLM_PID" 2>/dev/null || true
}
trap cleanup EXIT

echo "Waiting for vLLM to become healthy..."
until curl -sf "http://localhost:${VLLM_PORT}/health" > /dev/null 2>&1; do
    if ! kill -0 "$VLLM_PID" 2>/dev/null; then
        echo "ERROR: vLLM exited during startup. Tail of $VLLM_LOG:"
        tail -n 60 "$VLLM_LOG"
        exit 1
    fi
    sleep 3
done
echo "vLLM ready."

rc=0

if [ "$REQUEST_MODE" = "online" ] || [ "$REQUEST_MODE" = "both" ]; then
    echo "=== ONLINE path: $NUM_SAMPLES request(s) via extract_hidden_states_smoke.py ==="
    python scripts/extract_hidden_states_smoke.py \
        --endpoint "http://localhost:${VLLM_PORT}/v1" \
        --preprocessed-data "$DATA" \
        --num-samples "$NUM_SAMPLES" \
        --start-index "$START_INDEX" || rc=$?
fi

if [ "$REQUEST_MODE" = "offline" ] || [ "$REQUEST_MODE" = "both" ]; then
    echo "=== OFFLINE path: $NUM_SAMPLES sample(s) via data_generation_offline.py ==="
    rm -rf "$OFFLINE_OUT"
    python scripts/data_generation_offline.py \
        --endpoint "http://localhost:${VLLM_PORT}/v1" \
        --preprocessed-data "$DATA" \
        --output "$OFFLINE_OUT" \
        --max-samples "$NUM_SAMPLES" \
        --concurrency 1 \
        --validate-outputs \
        --fail-on-error || rc=$?
fi

if [ "$rc" -eq 0 ]; then
    echo "=== Hidden-state extraction smoke test PASSED ==="
else
    echo "=== Hidden-state extraction smoke test FAILED (rc=$rc). See $VLLM_LOG ==="
fi
exit "$rc"

=== extract_hidden_states_smoke.py ===
#!/usr/bin/env python3
"""Smoke test for vLLM hidden-state extraction, isolated from training.

This script does *not* launch vLLM. It expects a vLLM server already started
via ``scripts/launch_vllm.py`` (i.e. configured with the
``ExampleHiddenStatesConnector`` + ``extract_hidden_states`` speculative
config). It submits one (or a few) completion request(s), loads the extracted
hidden states from disk, and validates them exactly the way the offline/online
data pipelines do (``check_hidden_states``).

The point is to isolate hidden-state EXTRACTION from speculator training so we
can verify it works -- or pinpoint where it breaks -- for a given verifier
model, in particular hybrid-attention models (e.g. Qwen3.5/3.6-A3B).

It uses the exact same client call (``generate_hidden_states``) that the online
training data loader uses, so a pass here means the online extraction path is
healthy for the tested sample(s). For the offline path, point
``data_generation_offline.py`` at the same server.

Usage:
    python scripts/extract_hidden_states_smoke.py \
        --endpoint http://localhost:8000/v1 \
        --preprocessed-data ./output/<prepared_dataset> \
        --num-samples 1
"""

import argparse
import logging
import sys
from pathlib import Path

import openai
import torch
from datasets import load_from_disk
from safetensors.torch import load_file

from speculators.data_generation.offline import check_hidden_states
from speculators.data_generation.vllm_client import (
    DEFAULT_REQUEST_TIMEOUT,
    generate_hidden_states,
    wait_for_lock,
)
from speculators.train.data import build_client_item

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s %(levelname)s [hs_smoke] %(message)s"
)
logger = logging.getLogger("hs_smoke")


def parse_args():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--endpoint",
        default="http://localhost:8000/v1",
        help="vLLM endpoint configured for hidden-states extraction.",
    )
    parser.add_argument(
        "--preprocessed-data",
        required=True,
        help="Path to a prepare_data.py dataset; uses its 'input_ids' column.",
    )
    parser.add_argument(
        "--num-samples",
        type=int,
        default=1,
        help="Number of consecutive samples to submit (default: 1).",
    )
    parser.add_argument(
        "--start-index",
        type=int,
        default=0,
        help="First dataset index to submit (default: 0).",
    )
    parser.add_argument(
        "--request-timeout",
        type=float,
        default=DEFAULT_REQUEST_TIMEOUT,
        help="Per-request timeout in seconds.",
    )
    parser.add_argument(
        "--model",
        default=None,
        help="Optional model id; defaults to the server's first served model.",
    )
    parser.add_argument(
        "--keep-files",
        action="store_true",
        help="Do not delete the extracted safetensors files after validation.",
    )
    return parser.parse_args()


def describe(hs: torch.Tensor) -> str:
    f = hs.detach().float()
    return (
        f"shape={tuple(hs.shape)} dtype={hs.dtype} "
        f"min={f.min().item():.4f} max={f.max().item():.4f} "
        f"mean={f.mean().item():.4f} std={f.std().item():.4f} "
        f"nan={int(hs.isnan().sum())} inf={int(hs.isinf().sum())}"
    )


def main() -> int:
    args = parse_args()

    dataset = load_from_disk(args.preprocessed_data)
    dataset = dataset.with_format(
        "torch", columns=["input_ids"], output_all_columns=True
    )

    client = openai.Client(base_url=args.endpoint, api_key="EMPTY", max_retries=0)
    model = args.model or client.models.list().data[0].id
    logger.info("model=%s endpoint=%s dataset_size=%d", model, args.endpoint, len(dataset))

    end = min(args.start_index + args.num_samples, len(dataset))
    failures = 0
    for i in range(args.start_index, end):
        item = build_client_item(dataset[i])
        tokens = item["input_ids"]
        logger.info("[sample %d] submitting num_tokens=%d", i, len(tokens))
        try:
            path = generate_hidden_states(
                client, model, item, timeout=args.request_timeout
            )
            lock_path = path + ".lock"
            if Path(lock_path).exists():
                wait_for_lock(lock_path)

            data = load_file(path)
            check_hidden_states(data, tokens)
            logger.info(
                "[sample %d] OK token_ids=%d hidden_states %s",
                i,
                data["token_ids"].shape[0],
                describe(data["hidden_states"]),
            )
            if not args.keep_files:
                Path(path).unlink(missing_ok=True)
        except Exception as e:  # noqa: BLE001
            failures += 1
            logger.exception("[sample %d] FAILED: %s", i, e)

    total = end - args.start_index
    logger.info("Done: %d/%d samples succeeded", total - failures, total)
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())

```

### Test Result
Before (current `main`):
- Connector resolves `block_size=528` (global) vs buffer block size `22`.
- Hidden states are entirely zero (`min=max=mean=std=0`); fails validation, and
  OOB / `IndexKernel.cu:111 index out of bounds` at scale.

After (this branch):
- Connector resolves `block_size=22` on **both** worker and scheduler
  (`role=WORKER ... block_size=22`, `role=SCHEDULER ... block_size=22`).
- **200/200** samples succeed (online + offline) with healthy hidden states
  (`min≈-29.6, max≈43.5, mean≈0.001, std≈0.21, nan=0, inf=0`); no OOB, no OOM.

### CODE REVIEW

**Review by fynnsu — APPROVED** [review]
Generally looks good. I'm currently running some tests to make sure this does fix an issue we've seen in CI, will update once done.

**Comment by fynnsu** [comment]
> I'm currently running some tests to make sure this does fix an issue we've seen in CI, will update once done.

Yes, looks like this resolved the CI issue.

**Comment by imargulis** [comment]
@mgoin Thanks a lot for cleaning up the PR.
There are several random buildkite failures though

**Review by mgoin — APPROVED** [review]
Thanks, LGTM. Will force merge once Helen or Fynn retest to confirm
